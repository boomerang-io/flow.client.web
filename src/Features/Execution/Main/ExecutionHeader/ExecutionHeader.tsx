import React from "react";
import { useAppContext } from "Hooks";
import { useMutation, useQueryClient, UseQueryResult } from "react-query";
import { withRouter, Link, useParams } from "react-router-dom";
import CopyToClipboard from "react-copy-to-clipboard";
import moment from "moment";
import {
  Breadcrumb,
  BreadcrumbItem,
  Button,
  ComposedModal,
  ConfirmModal,
  FeatureHeader as Header,
  FeatureHeaderTitle as HeaderTitle,
  ModalBody,
  notify,
  SkeletonPlaceholder,
  Tag,
  TextArea,
  ToastNotification,
  TooltipHover,
} from "@boomerang-io/carbon-addons-boomerang-react";
import OutputPropertiesLog from "Features/Execution/Main/ExecutionTaskLog/TaskItem/OutputPropertiesLog";
import ErrorModal from "Components/ErrorModal";
import { appLink } from "Config/appConfig";
import { elevatedUserRoles, QueryStatus, ExecutionStatus } from "Constants";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { Catalog16, CopyFile16, StopOutline16, Warning16 } from "@carbon/icons-react";
import { WorkflowSummary } from "Types";
import styles from "./executionHeader.module.scss";

type Props = {
  history: any;
  location: any;
  match: any;
  workflow: UseQueryResult<any, Error> | UseQueryResult<any, Error> | UseQueryResult<any>;
  workflowExecution: UseQueryResult<any, Error> | UseQueryResult<any, Error> | UseQueryResult<any>;
  version: number;
};

const cancelSatusTypes = [ExecutionStatus.NotStarted, ExecutionStatus.Waiting, ExecutionStatus.InProgress];

function ExecutionHeader({ history, workflow, workflowExecution, version }: Props) {
  const { state } = history.location;
  const { user } = useAppContext();
  const queryClient = useQueryClient();

  const { platformRole } = user;
  const systemWorkflowsEnabled = elevatedUserRoles.includes(platformRole);
  const { teamName, initiatedByUserName, trigger, creationDate, scope, status, id } = workflowExecution.data;
  const displayCancelButton = cancelSatusTypes.includes(status);

  const { mutateAsync: deleteCancelWorkflowMutation } = useMutation(resolver.deleteCancelWorkflow, {
    onSuccess: () => queryClient.invalidateQueries(serviceUrl.getWorkflowExecution({ executionId: id })),
  });

  const handleCancelWorkflow = async () => {
    try {
      await deleteCancelWorkflowMutation({ executionId: id });
      notify(<ToastNotification kind="success" title="Cancel run" subtitle="Execution successfully cancelled" />);
    } catch {
      notify(<ToastNotification kind="error" title="Something's wrong" subtitle={`Failed to cancel this execution`} />);
    }
  };
  return (
    <Header
      className={styles.container}
      nav={
        <div className={styles.headerNav}>
          <Breadcrumb noTrailingSlash>
            <BreadcrumbItem>
              <Link to={state ? state.fromUrl : appLink.activity()}>{state ? state.fromText : "Activity"}</Link>
            </BreadcrumbItem>
            <BreadcrumbItem isCurrentPage>
              {!workflow?.data?.name ? (
                <SkeletonPlaceholder className={styles.workflowNameSkeleton} />
              ) : (
                <p>{workflow.data.name}</p>
              )}
            </BreadcrumbItem>
          </Breadcrumb>
          {workflow?.data && (
            <ComposedModal
              composedModalProps={{ shouldCloseOnOverlayClick: true }}
              modalHeaderProps={{
                title: "Advanced Detail",
                subtitle:
                  "Use the following to dive deeper and debug the execution. Tip: copy the commands into your local terminal and add the namespace.",
              }}
              modalTrigger={({ openModal }: { openModal: () => void }) => (
                <TooltipHover direction="right" content="Advanced Detail">
                  <button className={styles.workflowAdvancedDetailTrigger} onClick={openModal}>
                    <Catalog16 />
                  </button>
                </TooltipHover>
              )}
            >
              {() => <WorkflowAdvancedDetail workflow={workflow.data} />}
            </ComposedModal>
          )}
        </div>
      }
      header={
        <div style={{ display: "flex" }}>
          <HeaderTitle>Workflow run detail</HeaderTitle>
          {Boolean(workflowExecution?.data?.error?.code) && (
            <ComposedModal
              composedModalProps={{ shouldCloseOnOverlayClick: true }}
              modalHeaderProps={{ title: "Execution Error" }}
              modalTrigger={({ openModal }: { openModal: () => void }) => (
                <Button
                  className={styles.workflowErrorTrigger}
                  kind={"ghost"}
                  onClick={openModal}
                  renderIcon={Warning16}
                  size="small"
                >
                  View Execution Error
                </Button>
              )}
            >
              {() => (
                <ErrorModal
                  errorCode={workflowExecution?.data?.error?.code}
                  errorMessage={workflowExecution?.data?.error?.message ?? ""}
                />
              )}
            </ComposedModal>
          )}
        </div>
      }
      actions={
        workflowExecution.status === QueryStatus.Success ? (
          <div className={styles.content}>
            {workflowExecution.data.outputProperties &&
              Object.keys(workflowExecution.data.outputProperties).length > 0 && (
                <div className={styles.workflowOutputLog}>
                  <OutputPropertiesLog
                    isOutput
                    flowTaskName={workflow.data.name}
                    //@ts-ignore
                    flowTaskOutputs={workflowExecution.data.outputProperties}
                  />
                </div>
              )}
            {systemWorkflowsEnabled && (
              <dl className={styles.data}>
                <dt className={styles.dataTitle}>Scope</dt>
                <dd className={styles.dataValue}>{scope ?? "---"}</dd>
              </dl>
            )}
            <dl className={styles.data}>
              <dt className={styles.dataTitle}>Team</dt>
              <dd className={styles.dataValue}>{teamName ?? "---"}</dd>
            </dl>
            <dl className={styles.data}>
              <dt className={styles.dataTitle}>Version</dt>
              <dd className={styles.dataValue}>{version ?? "---"}</dd>
            </dl>
            <dl className={styles.data}>
              <dt className={styles.dataTitle}>Initiated by</dt>
              {initiatedByUserName ? (
                <dd className={styles.dataValue}>{initiatedByUserName}</dd>
              ) : (
                <dd aria-label="robot" aria-hidden={false} role="img">
                  {"ðŸ¤–"}
                </dd>
              )}
            </dl>
            <dl className={styles.data}>
              <dt className={styles.dataTitle}>Trigger</dt>
              <dd className={styles.dataValue}>{trigger}</dd>
            </dl>
            <dl className={styles.data}>
              <dt className={styles.dataTitle}>Start time</dt>
              <dd className={styles.dataValue}>{moment(creationDate).format("YYYY-MM-DD hh:mm A")}</dd>
            </dl>
            <dl className={styles.dataButton}>
              {displayCancelButton && (
                <ConfirmModal
                  affirmativeAction={handleCancelWorkflow}
                  affirmativeButtonProps={{ kind: "danger" }}
                  children="Are you sure? Once a workflow is cancelled it will stop executing."
                  title="Cancel run"
                  modalTrigger={({ openModal }: { openModal: () => void }) => (
                    <Button
                      className={styles.cancelRun}
                      data-testid="cancel-run"
                      kind="danger--ghost"
                      iconDescription="Cancel run"
                      onClick={openModal}
                      renderIcon={StopOutline16}
                      size="small"
                    >
                      Cancel run
                    </Button>
                  )}
                />
              )}
            </dl>
          </div>
        ) : (
          <SkeletonPlaceholder className={styles.headerContentSkeleton} />
        )
      }
    />
  );
}

function WorkflowAdvancedDetail({ workflow }: { workflow: WorkflowSummary }) {
  const { workflowId, executionId }: { workflowId: string; executionId: string } = useParams();
  const [copyTokenText, setCopyTokenText] = React.useState("Copy");

  const labelTexts = [`boomerang.io/workflow-id=${workflowId}`, `boomerang.io/workflow-activity-id=${executionId}`];

  if (Array.isArray(workflow.labels) && workflow.labels.length > 0) {
    workflow.labels.forEach((label) => {
      labelTexts.push(`${label.key}=${label.value}`);
    });
  }

  const kubernetesCommand = `kubectl get pods -l ${labelTexts.join(",")}`;
  const tektonCommand = `tkn tr list --label ${labelTexts.join(",")}`;

  return (
    <ModalBody>
      <div>Use this information to debug the execution using the Tekton CLI.</div>
      <h1 className={styles.detailHeading} style={{ marginTop: "0rem" }}>
        Labels
      </h1>
      <div className={styles.workflowLabels}>
        {labelTexts.map((label, index) => (
          <Tag key={`${label}-${index}`} className={styles.workflowLabelBubble} type="teal">
            {label}
          </Tag>
        ))}
      </div>
      <h1 className={styles.detailHeading}>Tekton Information</h1>
      <div>Use this information to debug the execution using the Tekton CLI.</div>
      <div className={styles.kubernetes}>
        <TextArea readOnly value={tektonCommand} />
        <TooltipHover direction="top" content={copyTokenText} hideOnClick={false}>
          <div className={styles.kubernetesCopyContainer}>
            <CopyToClipboard text={tektonCommand}>
              <Button
                className={styles.kubernetesCopy}
                iconDescription="copy-kubernetes"
                kind="ghost"
                onClick={() => setCopyTokenText("Copied!")}
                onMouseLeave={() => setCopyTokenText("Copy")}
                renderIcon={CopyFile16}
                size="small"
              />
            </CopyToClipboard>
          </div>
        </TooltipHover>
      </div>
      <h1 className={styles.detailHeading}>Kubernetes Information</h1>
      <div>Use this information to debug the execution using the Kubernetes CLI.</div>
      <div className={styles.kubernetes}>
        <TextArea readOnly value={kubernetesCommand} />
        <TooltipHover direction="top" content={copyTokenText} hideOnClick={false}>
          <div className={styles.kubernetesCopyContainer}>
            <CopyToClipboard text={kubernetesCommand}>
              <Button
                className={styles.kubernetesCopy}
                iconDescription="copy-kubernetes"
                kind="ghost"
                onClick={() => setCopyTokenText("Copied!")}
                onMouseLeave={() => setCopyTokenText("Copy")}
                renderIcon={CopyFile16}
                size="small"
              />
            </CopyToClipboard>
          </div>
        </TooltipHover>
      </div>
    </ModalBody>
  );
}

export default withRouter(ExecutionHeader);
