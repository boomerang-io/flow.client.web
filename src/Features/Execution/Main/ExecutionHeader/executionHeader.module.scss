@import "Styles/mixins";

.container {
  h1 {
    margin: 0;
  }
}

.headerNav {
  display: flex;
}

.workflowNameSkeleton {
  height: 1.125rem;
}

.workflowAdvancedDetailTrigger {
  align-items: center;
  cursor: pointer;
  display: flex !important;
  justify-content: center;
  &:focus {
    @include focus-outline();
  }
}

.workflowErrorTrigger {
  align-items: center;
  cursor: pointer;
  display: flex !important;
  justify-content: center;
  margin-left: 1rem;
}

.content {
  align-items: center;
  display: flex;
  margin: 1.125rem 0 1.5rem 0;
  overflow: hidden;
}

.data {
  color: var(--cds-text-01);
  margin-left: 2.5rem;
}

.dataButton {
  color: var(--cds-text-01);
  margin-left: 1.5rem;
}

.dataTitle {
  font-size: 0.875rem;
  font-weight: 300;
  letter-spacing: 0;
  line-height: 1.125rem;
  margin-bottom: 0.25rem;
}

.dataValue {
  font-size: 1rem;
  letter-spacing: 0;
  line-height: 1.25rem;
  max-width: 20rem;
  text-transform: capitalize;
  @include text-overflow-ellipsis();
}

.headerContentSkeleton {
  height: 4.31rem;
  width: 40rem;
}

.workflowOutputLog {
  margin-left: 0.5rem;
}

/**
 * AdvancedDetail Modal
 */

.detailHeading {
  margin: 1rem 0 0.5rem 0;
}

.workflowLabels {
  display: flex;
  flex-wrap: wrap;
}

.workflowLabelBubble {
  margin-left: 0;
}

.kubernetes {
  align-items: center;
  display: flex;
  margin-top: 1rem;
  padding-right: 1rem;
}

.kubernetesCopyContainer {
  height: 2rem;
  margin-left: 0.5rem;
  width: 2rem;
}

.kubernetesCopy {
  justify-content: center;
  width: 2rem;

  :global .bx--btn__icon {
    margin: 0;
  }
}

.cancelRun {
  margin-bottom: -1rem;
}