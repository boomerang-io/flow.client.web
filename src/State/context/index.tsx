import React from "react";
import { FlowTeam, FlowUser, TaskModel, WorkflowRevision, WorkflowSummary } from "Types";

export function createContext<ContextType>() {
  const context = React.createContext<ContextType | undefined>(undefined);
  function useContext() {
    const contextValue = React.useContext(context);
    if (!contextValue) throw new Error("useContext must be inside a Provider with a value");
    return contextValue;
  }
  return [useContext, context.Provider] as const;
}

export const [useAppContext, AppContextProvider] = createContext<AppContext>();

type AppContext = {
  isTutorialActive: boolean;
  communityUrl: string;
  setIsTutorialActive: (isActive: boolean) => void;
  user: FlowUser;
  teams: FlowTeam[];
};

interface taskProvider {
  category: string;
  id: string;
  icon: any;
  name: string;
}

interface stepInterface {
  activityId: string;
  duration: number;
  flowTaskStatus: string;
  id: string;
  order: number;
  outputs: any;
  startTime: string;
  taskId: string;
  taskName: string;
  runWorkflowActivityStatus: string;
}

interface workflowExecutionInterface {
  creationDate: string;
  duration: number;
  id: string;
  properties: any; //array, may need to be changed later
  status: string;
  steps: Array<stepInterface>;
  teamName: string;
  trigger: string;
  workflowId: string;
  workflowRevisionid: string;
}

interface ExecutionContext {
  tasks: Array<taskProvider>;
  workflowExecution: workflowExecutionInterface;
  workflowRevision: object;
}
export const [useExecutionContext, ExecutionContextProvider] = createContext<ExecutionContext>();

interface EditorContext {
  revisionDispatch?: Function;
  revisionState: WorkflowRevision;
  summaryData: WorkflowSummary;
  taskTemplatesData: Array<TaskModel>;
}

export const [useEditorContext, EditorContextProvider] = createContext<EditorContext>();
