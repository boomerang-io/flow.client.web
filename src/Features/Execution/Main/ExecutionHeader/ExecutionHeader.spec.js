import React from "react";
import ExecutionHeader from "./index";
import { teams, profile } from "ApiServer/fixtures";
import { AppContextProvider } from "State/context";
const props = {
  workflowExecution: {
    version: 1,
    status: "success",
    data: {
      teamName: "CAI Offerings",
      initiatedByUserName: "Tim Bula",
      trigger: "manual",
      creationDate: "2019-09-03T15:00:00.049+0000",
    },
  },
  workflow: {
    isFetching: false,
    data: {
      name: "Sparkle Flow with extra glitter and donuts on the side",
    },
  },
};

describe("ExecutionHeader --- Snapshot", () => {
  it("Capturing Snapshot of ExecutionHeader", () => {
    const { baseElement } = rtlRouterRender(
      <AppContextProvider
        value={{
          isTutorialActive: false,
          setIsTutorialActive: () => {},
          user: profile,
          teams,
        }}
      >
        <ExecutionHeader {...props} />
      </AppContextProvider>
    );
    expect(baseElement).toMatchSnapshot();
  });
});
