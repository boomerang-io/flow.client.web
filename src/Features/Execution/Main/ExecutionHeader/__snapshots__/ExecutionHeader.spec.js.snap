// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionHeader --- Snapshot Capturing Snapshot of ExecutionHeader 1`] = `
<body
  id="app"
>
  <div>
    <header
      class="bx--bmrg-feature-header container --bordered"
    >
      <section
        class="bx--bmrg-feature-header__content"
      >
        <div
          class="bx--bmrg-feature-header__nav"
        >
          <div
            class="headerNav"
          >
            <nav
              aria-label="Breadcrumb"
            >
              <ol
                class="bx--breadcrumb bx--breadcrumb--no-trailing-slash"
              >
                <li
                  class="bx--breadcrumb-item"
                >
                  <a
                    class="bx--link"
                    href="/activity"
                  >
                    Activity
                  </a>
                </li>
                <li
                  class="bx--breadcrumb-item bx--breadcrumb-item--current"
                >
                  <p
                    class="bx--link"
                  >
                    Sparkle Flow with extra glitter and donuts on the side
                  </p>
                </li>
              </ol>
            </nav>
            <button
              class="workflowAdvancedDetailTrigger"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M26,2H8A2,2,0,0,0,6,4V8H4v2H6v5H4v2H6v5H4v2H6v4a2,2,0,0,0,2,2H26a2,2,0,0,0,2-2V4A2,2,0,0,0,26,2Zm0,26H8V24h2V22H8V17h2V15H8V10h2V8H8V4H26Z"
                />
                <path
                  d="M14 8H22V10H14zM14 15H22V17H14zM14 22H22V24H14z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="bx--bmrg-feature-header__header"
        >
          <div
            style="display: flex;"
          >
            <h1
              class="bx--bmrg-feature-header-title"
            >
              Workflow run detail
            </h1>
          </div>
        </div>
      </section>
      <div
        class="content"
      >
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Scope
          </dt>
          <dd
            class="dataValue"
          >
            ---
          </dd>
        </dl>
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Team
          </dt>
          <dd
            class="dataValue"
          >
            CAI Offerings
          </dd>
        </dl>
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Version
          </dt>
          <dd
            class="dataValue"
          >
            ---
          </dd>
        </dl>
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Initiated by
          </dt>
          <dd
            class="dataValue"
          >
            Tim Bula
          </dd>
        </dl>
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Trigger
          </dt>
          <dd
            class="dataValue"
          >
            manual
          </dd>
        </dl>
        <dl
          class="data"
        >
          <dt
            class="dataTitle"
          >
            Start time
          </dt>
          <dd
            class="dataValue"
          >
            2020-01-01 12:00 AM
          </dd>
        </dl>
        <dl
          class="dataButton"
        />
      </div>
    </header>
  </div>
  <div
    class="bx--bmrg-modal-portal"
  />
</body>
`;
